base/leaky_bucket.go

```diff
@@ -3,34 +3,42 @@ package base
 import (
 	"errors"
 	"fmt"
+	"sync/atomic"
 	"time"
 
+	"github.com/couchbase/gocb"
 	sgbucket "github.com/couchbase/sg-bucket"
+	pkgerrors "github.com/pkg/errors"
 )
 
 // A wrapper around a Bucket to support forced errors.  For testing use only.
 type LeakyBucket struct {
-	bucket    Bucket
-	incrCount uint16
-	config    LeakyBucketConfig
+	bucket       Bucket
+	incrCount    uint16
+	config       LeakyBucketConfig
+	numBucketOps uint64 // Track the total number of bucket ops
 }
 
 // The config object that controls the LeakyBucket behavior
 type LeakyBucketConfig struct {
 
 	// Incr() fails 3 times before finally succeeding
-	IncrTemporaryFailCount uint16
+	IncrTemporaryFailCount uint16 `json:"incr_temporary_fail_count,omitempty"`
 
 	// Emulate TAP/DCP feed de-dupliation behavior, such that within a
 	// window of # of mutations or a timeout, mutations for a given document
 	// will be filtered such that only the _latest_ mutation will make it through.
-	TapFeedDeDuplication bool
-	TapFeedVbuckets      bool     // Emulate vbucket numbers on feed
+	TapFeedDeDuplication bool     `json:"tap_feed_deduplication,omitempty"`
+	TapFeedVbuckets      bool     `json:"tap_feed_vbuckets,omitempty"` // Emulate vbucket numbers on feed
 	TapFeedMissingDocs   []string // Emulate entry not appearing on tap feed
 
 	// Returns a partial error the first time ViewCustom is called
-	FirstTimeViewCustomPartialError bool
+	FirstTimeViewCustomPartialError bool                                                       `json:"first_time_view_custom_partial_error,omitempty"`
 	PostQueryCallback               func(ddoc, viewName string, params map[string]interface{}) // Issues callback after issuing query when bucket.ViewQuery is called
+
+	// Every Nth bucket op will emulate a temporary/recoverable Couchbase Server errors.  0 means no errors.  This should
+	// be set to a number high enough to get past the startup ops.  Currently, that is > 3.
+	InjectNthBucketOpArtificialErrors uint `json:"inject_nth_bucket_op_artificial_errors"`
```

Comment:
I find this property name a bit unintuitive.  What do you think about something like 'bucket_503_error_frequency'?

Reply:
How about `bucket_503_injected_error_frequency`?  The "injected" just makes it explicit that this setting is about injecting errors, as opposed to handling/logging/etc around those errors.